cirq-core/cirq/transformers/gauge_compiling/gauge_compiling.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Creates the abstraction for gauge compiling as a cirq transformer."""
+
+from typing import Callable, Tuple, Optional, Sequence, Union, List
+import abc
+import itertools
+import functools
+from dataclasses import dataclass
+
+import numpy as np
+
+from cirq.transformers import transformer_api
+from cirq import ops, circuits
+
+
+_SINGLE_QUBIT_GATES_T = Union[ops.Gate, Sequence[ops.Gate]]
+
+
+class Gauge(abc.ABC):
+    """A gauge replaces a two qubit gate with an equivalent subcircuit.
+    0: pre_q0───────two_qubit_gate───────post_q0
+                        |
+    1: pre_q1───────two_qubit_gate───────post_q1
+
+    The Gauge class in general represents a family of closely related gauges
+    (e.g. random z-rotations); Use `sample` method to get a specific gauge.
+    """
+
+    def weight(self) -> float:
+        """Returns the relative frequency for selecting this gauge."""
+        return 1.0
+
+    @abc.abstractmethod
+    def sample(self, gate: ops.Gate, prng: np.random.Generator) -> "ConstantGauge":
+        """Returns a ConstantGauge sampled from a family of gauges.
+
+        Args:
+            gate: The two qubit gate to replace.
+            prng: A numpy random number generator.
+
+        Returns:
+            A ConstantGauge.
+        """
+
+
+@dataclass(frozen=True)
+class ConstantGauge(Gauge):
+    """A gauge that replaces a two qubit gate with a constant gauge."""
+
+    two_qubit_gate: ops.Gate
+    pre_q0: _SINGLE_QUBIT_GATES_T = ()
+    pre_q1: _SINGLE_QUBIT_GATES_T = ()
+    post_q0: _SINGLE_QUBIT_GATES_T = ()
+    post_q1: _SINGLE_QUBIT_GATES_T = ()
+
+    def sample(self, gate: ops.Gate, prng: np.random.Generator) -> "ConstantGauge":
+        return self
+
+    @functools.cached_property
+    def pre(self) -> Tuple[Sequence[ops.Gate], Sequence[ops.Gate]]:
+        """A tuple (ops to apply to q0, ops to apply to q1)."""
+        return _as_sequence(self.pre_q0), _as_sequence(self.pre_q1)
+
+    @functools.cached_property
+    def post(self) -> Tuple[Sequence[ops.Gate], Sequence[ops.Gate]]:
+        """A tuple (ops to apply to q0, ops to apply to q1)."""
+        return _as_sequence(self.post_q0), _as_sequence(self.post_q1)
+
+
+def _select(choices: Sequence[Gauge], probabilites: np.ndarray, prng: np.random.Generator) -> Gauge:
+    return choices[prng.choice(len(choices), p=probabilites)]
```

Comment:
The `_select` is only used here - consider moving that line to `GaugeSelector.__call__`.

Reply:
it's needed here in order to force select specific gauges in tests